### PR TITLE
CRM-14987 find correct WP install path

### PR DESCRIPTION
**Pull request by narayanis**
No errors in my testing, and should fix CRM-14987 as well as recently-reported open_basedir() security warnings. I took the liberty of defining the constant if it's not found in civicrm.settings.php; please advise if that's too presumptuous.

---
- [CRM-14987: WordPress cmsRootPath can incorrectly return unrelated CMS install](https://issues.civicrm.org/jira/browse/CRM-14987)

### CODE REVIEW

**Comment by colemanw** [comment]
jenkins, ok to test

**Comment by kcristiano** [comment]
@colemanw This PR is related:  https://github.com/civicrm/civicrm-core/pull/5749  I had originally done this against master, but as we wanted it in 4.6 it was redone and committed to 4.6 branch.   As far as I can tell, this patch has not made it to master.   I do believe this could be a duplication of that.   Do I need to rebase 5749 to get it in master?    I had thought that these would be upstreamed.   4.6 PRs that got committed are:

https://github.com/civicrm/civicrm-core/pull/5833
https://github.com/civicrm/civicrm-core/pull/5834
related:  https://github.com/civicrm/civicrm-wordpress/pull/75

**Comment by colemanw** [comment]
@kcristiano 4.6 gets forward-merged into master on a regular basis. It looks to me like the commits are in master. See https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Permission/WordPress.php#L65

**Comment by kcristiano** [comment]
@colemanw Thank You -- I see them now.   My local master did not have them.   I'll rebuild locally.   I was more concerned about https://github.com/civicrm/civicrm-core/blob/master/CRM/Utils/System/WordPress.php#L345  and wanted to be sure that was in before this was added.

**Comment by narayanis** [comment]
We may be trying to solve the same problem (just showing as different symptons) in multiple ways. When using shared hosting, if CIVICRM_CMSDIR isn't defined, the method of exploding and parsing each level to find the wp home path throws security errors.  My commit uses the WP function get_home_path() to let WordPress do that work for us. If this is a setting in the Civi DB now, do we still need the CIVICRM_CMSDIR constant or the validInstallDir() function at all? The only thing that function does is arbitrarily check for wp-includes/version.php. Is a bigger refactor to this file needed?

**Comment by colemanw** [comment]
@christianwach might know more about this area

**Comment by christianwach** [comment]
I've noticed the multiple reports of `open_basedir` issues and agree that rewriting these method for WordPress should fix the problem. But - in my opinion - this PR doesn't address the fact that we *already have* the canonical path to WordPress. Therefore `CRM_Utils_System_WordPress::cmsRootPath()` can simply be:

```php
public function cmsRootPath() {
  $path = CRM_Core_BAO_Setting::getItem('CiviCRM Preferences', 'wpLoadPhp');
  if (file_exists($path)) {
    return $path;
  }
  return NULL;
}```

This makes `CRM_Utils_System_WordPress::validInstallDir()` basically redundant, though (if it is used elsewhere) it could be rewritten to do the `file_exists()` check for the `$path` variable as above.

**Comment by christianwach** [comment]
Doh. Hold on. `$path` is not the same as `ABSPATH`, of course. It just needs 'wp-load.php' stripping from it.

**Comment by narayanis** [comment]
A search of the repo only found `validInstallDir()` in this one file, so the code already has multiple paths to reach similar ends. They are:
* A Civi constant (optional) called `CIVICRM_CMSDIR` only found within WordPress.php
* A new setting saved in the DB called `wpLoadPhp`, found in Core.setting.php and WordPress.php
* A function that explodes the path of the current script, then tries to guess its way to the WP folder
* Native WP constants. Based on what I see in install/index.php, we already assume that WP constants are available.

Do we need so many moving parts? Or can we rely on the existence of WP constants at all times (not just during install)? Also, please let me know if this convo should move over to the issue in JIRA or to a forum instead of here.

**Comment by christianwach** [comment]
> A Civi constant (optional) called CIVICRM_CMSDIR only found within WordPress.php

This looks pretty pointless to me. If it must remain, then it should be accommodated in `cmsRootPath()`, by returning it if it exists prior to returning the `wpLoadPhp` setting.

> A new setting saved in the DB called wpLoadPhp, found in Core.setting.php and WordPress.php

This is the most reliable way to find 'wp-load.php'.

> A function that explodes the path of the current script, then tries to guess its way to the WP folder

This should be ditched or rewritten, IMO.

> Native WP constants. Based on what I see in install/index.php, we already assume that WP constants are available.

The `wpLoadPhp` setting is saved after Civi has been installed - on the first page reload thereafter.

**Comment by narayanis** [comment]
The `CIVICRM_CMSDIR` constant has been around since at least 4.3. The function to ditch is what I want to address; it just opened a can of worms. The `wpLoadPhp` setting saves the full path to `wp-load.php`, leaving extra checking of the same code (depending on whether the setting is empty). What do you think of an approach like this?

```php
public function cmsRootPath() {
    $cmsRoot = $valid = NULL;
    if (defined('CIVICRM_CMSDIR')) {
      $cmsRoot = CIVICRM_CMSDIR;
      if (file_exists($cmsRoot)) {
        $valid = TRUE;
      }
    }
    
    if (!($valid)) {
      $cmsRoot = CRM_Core_BAO_Setting::getItem('CiviCRM Preferences', 'cmsRootPath');
      if (file_exists($cmsRoot)) {
        $valid = TRUE;
        }
    }
    return ($valid) ? $cmsRoot : NULL;
  }

public function loadBootStrap($name = NULL, $pass = NULL) {
    global $wp, $wp_rewrite, $wp_the_query, $wp_query, $wpdb;

    $cmsRootPath = $this->cmsRootPath();
    if (!$cmsRootPath) {
      CRM_Core_Error::fatal("Could not find the install directory for WordPress");
    }

    $wpLoadPath = $cmsRootPath . DIRECTORY_SEPARATOR . 'wp-load.php';
    if (file_exists($wpLoadPath)) {
      require_once $wpLoadPath;
    }
    else {
      CRM_Core_Error::fatal("Could not find the bootstrap file for WordPress");
    }

    $wpUserTimezone = get_option('timezone_string');
    if ($wpUserTimezone) {
      date_default_timezone_set($wpUserTimezone);
      CRM_Core_Config::singleton()->userSystem->setMySQLTimeZone();
    }

    $wpPluggablePath = $cmsRootPath . DIRECTORY_SEPARATOR . 'wp-includes/pluggable.php';
    if (file_exists($wpPluggablePath)) {
      require_once $wpPluggablePath
    }
    else {
      CRM_Core_Error::fatal("Could not find the pluggable file for WordPress");
    }

...
...

```

**Comment by christianwach** [comment]
A few observations:

* In `cmsRootPath()`, I'd imagine that `CIVICRM_CMSDIR` is not a path to a file, but to a directory.
* Also, the `wpLoadPhp` setting is already `ABSPATH . wp-load.php`, so you'll need to strip the filename with  `$cmsRoot = str_replace('wp-load.php', '', $cmsRoot)` inside the second `if (file_exists($cmsRoot))` of `cmsRootPath()`. Makes me think it was a mistake of mine to store the full path rather than a direct copy of `ABSPATH`. :blush: 
* In `loadBootStrap()`, the line `if (file_exists($wpLoadPath)) {` seems redundant,since we've already checked that in `cmsRootPath()`. It should just read `if(!is_null($wpLoadPath)) {`.
* Also in `loadBootStrap()`, none of those global variables are actually used. The line `global $wp, $wp_rewrite, $wp_the_query, $wp_query, $wpdb;` could therefore go.

Other than these, I'd say that's looking a lot better.

**Comment by kcristiano** [comment]
@narayanis @christianwach    I agree that this is looking good.   For compatibility reasons, would we want to leave the setting name as  `wpLoadphp`, but only store the path to wp-load.php?

**Comment by christianwach** [comment]
Ah yes, @kcristiano makes a good point. I missed that you're referencing a non-existent setting, @narayanis.

**Comment by christianwach** [comment]
@kcristiano I can add a `wpAbsPath` setting alongside `wpLoadPhp` if that makes life easier.

**Comment by christianwach** [comment]
The reason I didn't add a `wpAbsPath` setting is that as soon as WordPress is bootstrapped by including the wp-load.php file, all of the WordPress constants are immediately available. As indeed is all of WordPress.

tl;dr I'm not clear why Civi would need to know if it's in a valid install directory.

**Comment by kcristiano** [comment]
@christianwach I am just thinking about backwards compatibility.  If we add a `wpAbsPath` setting along side, we'd need to test for both or starting in 4.7 deprecate `wpLoadphp` and just not use it.   

I am thinking that changing the value in `wpLoadphp` might be the better way to go.

I'll review the install directory perspective, I have gone through all of the install files as part of that refactoring and with the new settings, that may not be needed.

**Comment by narayanis** [comment]
What about moving the error check to `wpAbsPath` within `cmsRootPath()`, and while we're at it, use ABSPATH for pluggable.php since it will be available by then?

```php
public function cmsRootPath() {
    $cmsRoot = $valid = NULL;

    //check for the constant from civicrm.settings.php first
    if (defined('CIVICRM_CMSDIR')) {
      $cmsRoot = CIVICRM_CMSDIR;
      if (file_exists($cmsRoot)) {
        $valid = TRUE;
      }
    }
    
    //check if we have the setting in the DB next
    if (!($valid)) {
      $cmsRoot = CRM_Core_BAO_Setting::getItem('CiviCRM Preferences', 'wpAbsPath');
      if (file_exists($cmsRoot)) {
        $valid = TRUE;
        }
    }

    if (!($valid)) {
      CRM_Core_Error::fatal("Could not find the install directory for WordPress");
    }
    else {
      return $cmsRoot;
    }
  }

public function loadBootStrap($name = NULL, $pass = NULL) {
    global $wp, $wp_rewrite, $wp_the_query, $wp_query, $wpdb;

    $cmsRootPath = $this->cmsRootPath();

    $wpLoadPath = $cmsRootPath . DIRECTORY_SEPARATOR . 'wp-load.php';
    if (file_exists($wpLoadPath)) {
      require_once $wpLoadPath;
    }
    else {
      CRM_Core_Error::fatal("Could not find the bootstrap file for WordPress");
    }

    $wpUserTimezone = get_option('timezone_string');
    if ($wpUserTimezone) {
      date_default_timezone_set($wpUserTimezone);
      CRM_Core_Config::singleton()->userSystem->setMySQLTimeZone();
    }

    require_once ABSPATH . DIRECTORY_SEPARATOR . 'pluggable.php';

...
...

``` 
BTW I originally the new setting `cmsRoot` to be a bit more generic, in case Drupal and Joomla versions could get some use out of it.

**Comment by totten** [comment]
I haven't been too familiar with this issue, but I chatted with @kcristiano about this PR. A few comments/observations:

 * I don't think we should remove the search. For existing installations with vanilla file hierarchy, it could continue to work without any special settings/defines.
 * It does make sense to try to read the path and save it. However, a better time to do this is during installation (when creating `civicrm.settings.php`): when you deal with dev/staging/production workflows, anything with an absolute path or URL needs to be adjusted. There's already several of these in `civicrm.settings.php`. (If we put the data in DB, then it's harder to admin to find/update.)

So, for example, this might mean:

1. Update the `civicrm.settings.php.template`
2. Update installer to fill in `CIVICRM_CMSDIR` using `get_home_path()`

**Comment by christianwach** [comment]
> I don't think we should remove the search. For existing installations with vanilla file hierarchy, it could continue to work without any special settings/defines.

@totten I don't follow why the search (if by that you're referring to the directory traversal that throws PHP warnings when `open_basedir` is in effect!) is necessary if the path is stored via the procedure below. In addition, there will be the `wpLoadPhp` setting that can be read in vanilla installs. Have you got a specific reason for retaining it?

> It does make sense to try to read the path and save it. However, a better time to do this is during installation (when creating civicrm.settings.php)

Agree entirely with this. It would be far better to save whatever WordPress paths and constants CiviCRM needs access to during the install process when it can be guaranteed that WordPress is bootstrapped.

**Comment by kcristiano** [comment]
@christianwach I believe the directory traversal is needed for existing installs that do not have CIVICRM_CMSDIR defined.   

We added a commented out setting in `civicrm.settings.php` next step is to populate that with the path to WP.  

I'll link to these comments in CRM-14987

**Comment by christianwach** [comment]
> I believe the directory traversal is needed for existing installs that do not have CIVICRM_CMSDIR defined.

@kcristiano I think I'm getting confused somewhere along the line on this issue. As I understand it, the install process takes place when WordPress is bootstrapped, so the paths are available as WordPress constants during that process. Once installed, `wpLoadPhp` is available, which gives the canonical path to 'wp-load.php' - and therefore to the WordPress install. I struggle to understand what the directory traversal adds - apart from lots of log entries when `open_basedir` is in effect!

**Comment by kcristiano** [comment]
@christianwach Upon reflection I am sure you are right.   My conservative approach was due to worrying that some Civi installs might not have the setting, but if we use `wpLoadPhp` for 4.7 on it should be safe as all 4.6 upgrades get the setting on first load.

This would also be simpler to implement than changing `civicrm.setting.php`.  I'll add these notes to CRM-14987.
